src/Domain/ShoppingItemStack.php

```diff
@@ -81,6 +81,26 @@ public function push(ShoppingItem $item): self
         return $copy;
     }
 
+    /**
+     * Delete/remove a shopping item.
+     *
+     * @param ShoppingItem $itemToRemove
+     * @return $this
+     */
+    public function remove(ShoppingItem $itemToRemove): self
+    {
+        $stack = $this->all();
+
+        $items = [];
+        foreach ($stack as $item) {
```

Comment:
Comment: here's some new knowledge for you.

Typically you can iterate over an object - i.e. if you do `foreach ($object as $value)` you'd get an error message about having an invalid value for foreach. However, there is a way of making objects iterable - if they implement the `Traversable` interface, they can be iterated over. Typically in objects we tend to use the `IteratorAggregate` interface - this extends `Traversable`, so any objects that use it can be iterated over.

If you take a look at `ShoppingItemStack`, you'll see that it implements the `IteratorAggregate` interface. Therefore at line 95 we can do:

```php
foreach ($this as $item) {
```

I.e. we are iterating over the current object. That means you can remove the `$stack = $this->all()` assignment, as there is no need to assign the stack when you can just loop over `$this` (the current object).

Reply:
Some docs for you to take a look at:
https://www.php.net/manual/en/class.traversable.php
https://www.php.net/manual/en/class.iteratoraggregate.php

You'll notice in `ShoppingItemStack::getIterator()` it uses the expression `yield from`. The `yield` keyword turns a function or method into a generator. This is quite a complex concept, so don't worry if it doesn't make a lot of sense now but you could take a look at the generator syntax here:
https://www.php.net/manual/en/language.generators.syntax.php

Generators exists in a lot of programming languages - e.g. Javascript also has generators.

Reply:
I did wonder if there was a way to iterate over an object but that does make sense!